### PR TITLE
Support smearing

### PR DESCRIPTION
**Pull request by antoine-levitt**
This is unit tested but I didn't try it on an actual metal. Also added an assert in the case of zero temperature that we are in an insulator (otherwise the procedure is not correct); that assert makes one test fail, do you know why that is?

### CODE REVIEW

**Comment by mfherbst** [comment]
Not sure. I'll look at it once I am done with the other energy computation and density computation things.

**Comment by mfherbst** [comment]
I did some digging: The answer is simple. The `scf_compare` routine uses a Hamiltonian without XC potential ... and at that level of theory silicon is a metal.

**Comment by antoine-levitt** [comment]
Oh interesting! For metals at 0K the routine is incorrect, I'll take care of it and remove the assert.

**Comment by mfherbst** [comment]
I'm not sure that's the right way to go. I think it's better to have an `occupation_step` function (or something like that) which plainly does no smearing and one with smearing. The user should then either give a temperature and select smearing or not.

**Comment by mfherbst** [comment]
@antoine-levitt Just pushed the changes as I had them in mind. Feel free to have a look.

**Comment by antoine-levitt** [comment]
I agree; I just meant that there is one routine missing, which is a metal at 0K, which is different than an insulator. This is just making sure that the T=0 case of `occupation_temperature` is handled appropriately. I think the `occupation_step` function should still raise an error when HOMO > LUMO, as in that case the result is just plain wrong. If you agree I'll revert your commit, fix `occupation_temperature` to accept T=0, and add an optional flag to disable the error checking.

**Comment by mfherbst** [comment]
I mainly agree ... the error flag has to be accessible from the tests, most importantly.

I do not want to find out right now, how to generate the reference data in ABINIT for silicon without XC, which is actually a metal.

**Comment by antoine-levitt** [comment]
Are you sure abinit gives the result of `occupation_step` (ie with a HOMO > LUMO)? That seems weird.

**Comment by mfherbst** [comment]
Yes:
- The values agree too good.
- The default of [`occopt`](https://docs.abinit.org/variables/basic/#occopt), which I did not change, puts 2 or 0 in each k-Point.

File: src/core/occupation.jl [hunk 1]
```diff
@@ -1,20 +1,79 @@
...
+H2(x) = 4x^2 - 2
+H3(x) = 8x^3 - 12x
+A_coeff(n) = (-1)^n / (factorial(n) * 4^n * sqrt(π))
+const A1 = A_coeff(1)
+const A2 = A_coeff(2);
+smearing_methfessel_paxton_1(x) = smearing_gaussian(x) + A1 * H1(x) * exp(-x^2)
+function smearing_methfessel_paxton_2(x)
+    smearing_gaussian(x) + A1 * H1(x) * exp(-x^2) + A2 * H3(x) * exp(-x^2)
+end
+
```
Comment (antoine-levitt) [inline src/core/occupation.jl]: I understand you don't like long lines, but in this case it kills off the symmetry with the previous line and makes it harder to see it's just one more term

File: src/utils/self_consistent_field.jl [hunk 2]
```diff
@@ -11,15 +11,20 @@ electrons, `ρ` is the initial density, e.g. constructed via a SAD guess.
...
-    if ρ === nothing
-        ρ = guess_hcore(ham, n_bands, compute_occupation, lobpcg_prec=lobpcg_prec)
+    if smearing === nothing
+        compute_occupation(basis, energies, Psi) = occupation_step(basis, energies, Psi,
+                                                                   n_electrons)
+    else
+        function compute_occupation(basis, energies, Psi)
+            occupation_temperature(basis, energies, Psi, n_electrons, T, smearing=smearing)
+        end
     end
```
Comment (antoine-levitt) [inline src/utils/self_consistent_field.jl]: Hm, I get 
`WARNING: Method definition (::getfield(DFTK, Symbol("#compute_occupation#188")){T, smearing, n_electrons})(Any, Any, Any) in module DFTK at /home/antoine/.julia/dev/DFTK/src/utils/self_consistent_field.jl:18 overwritten at /home/antoine/.julia/dev/DFTK/src/utils/self_consistent_field.jl:22.
`
from this

Comment (antoine-levitt) [reply src/utils/self_consistent_field.jl]: WTF: https://discourse.julialang.org/t/conditional-function-definition/27348

**Comment by antoine-levitt** [comment]
I left the check disabled by default for tests, because I don't have a good way of passing down the flag to the function, but this should really be enabled by default. The arguments to the scf function is starting to get huge, we should start thinking of putting it into a struct.

File: src/utils/self_consistent_field.jl [hunk 3]
```diff
@@ -14,17 +14,16 @@ function self_consistent_field(ham::Hamiltonian, n_bands::Int, n_electrons::Int;
                                ρ=nothing, tol=1e-6, T=0, smearing=nothing,
                                lobpcg_prec=PreconditionerKinetic(ham, α=0.1),
                                max_iter=100, algorithm=:scf_nlsolve, kwargs...)
-    if smearing === nothing
-        compute_occupation(basis, energies, Psi) = occupation_step(basis, energies, Psi,
-                                                                   n_electrons)
+    if smearing == nothing
```
Comment (mfherbst) [inline src/utils/self_consistent_field.jl]: Tests for nothing should have `===`, see e.g. https://docs.julialang.org/en/latest/base/base/#Base.isnothing

File: src/utils/self_consistent_field.jl (hunk 2 above)
Comment (mfherbst) [reply src/utils/self_consistent_field.jl]: Interesting ... weird one for sure.

File: src/core/occupation.jl (hunk 1 above)
Comment (mfherbst) [reply src/core/occupation.jl]: How about
```
smearing_methfessel_paxton_2(x) = smearing_methfessel_paxton_1(x) + A2 * H3(x) * exp(-x^2) 
```
That's shorter and makes it apparent, too?

File: src/utils/self_consistent_field.jl (hunk 3 above)
Comment (antoine-levitt) [reply src/utils/self_consistent_field.jl]: Hm, you're right. Should not matter in principle but the compiler is sure that `x === nothing` is false as long as the type of `x` is not Nothing, whereas `==` might do stranger things. Thanks!

**Comment by mfherbst** [comment]
LGTM as it is now.

**Pull request merged**